src/libraries/Hooks.sol

```diff
@@ -151,44 +193,103 @@ library Hooks {
     {
+        amountToSwap = params.amountSpecified;
+        if (msg.sender == address(self)) return (amountToSwap, hookDeltaSpecified);
         if (self.hasPermission(BEFORE_SWAP_FLAG)) {
-            self.callHook(abi.encodeWithSelector(IHooks.beforeSwap.selector, msg.sender, key, params, hookData));
+            hookDeltaSpecified = self.callHookWithReturnDelta(
+                abi.encodeWithSelector(IHooks.beforeSwap.selector, msg.sender, key, params, hookData),
+                self.hasPermission(BEFORE_SWAP_RETURNS_DELTA_FLAG)
+            ).toInt128();
+
+            // Update the swap amount according to the hook's return, and check that the swap type doesnt change (exact input/output)
+            if (hookDeltaSpecified != 0) {
+                bool exactInput = amountToSwap < 0;
+                amountToSwap += hookDeltaSpecified;
+                if (exactInput ? amountToSwap > 0 : amountToSwap < 0) revert HookDeltaExceedsSwapAmount();
+            }
         }
     }
 
     /// @notice calls afterSwap hook if permissioned and validates return value
+    /// @dev if noSelfCall bypasses the call, hookDeltaUnspecified will be 0
     function afterSwap(
         IHooks self,
         PoolKey memory key,
         IPoolManager.SwapParams memory params,
-        BalanceDelta delta,
-        bytes calldata hookData
-    ) internal noSelfCall(self) {
+        BalanceDelta swapDelta,
+        bytes calldata hookData,
+        int128 hookDeltaSpecified
+    ) internal returns (BalanceDelta swapperDelta, BalanceDelta hookDelta) {
+        if (msg.sender == address(self)) return (swapDelta, BalanceDeltaLibrary.ZERO_DELTA);
+
+        int128 hookDeltaUnspecified;
+        swapperDelta = swapDelta;
         if (self.hasPermission(AFTER_SWAP_FLAG)) {
-            self.callHook(abi.encodeWithSelector(IHooks.afterSwap.selector, msg.sender, key, params, delta, hookData));
+            hookDeltaUnspecified = self.callHookWithReturnDelta(
+                abi.encodeWithSelector(IHooks.afterSwap.selector, msg.sender, key, params, swapDelta, hookData),
+                self.hasPermission(AFTER_SWAP_RETURNS_DELTA_FLAG)
+            ).toInt128();
+        }
+
+        if (hookDeltaUnspecified != 0 || hookDeltaSpecified != 0) {
+            hookDelta = (params.amountSpecified < 0 == params.zeroForOne)
+                ? toBalanceDelta(hookDeltaSpecified, hookDeltaUnspecified)
+                : toBalanceDelta(hookDeltaUnspecified, hookDeltaSpecified);
+
+            // the caller has to pay for (or receive) to hook's delta
+            swapperDelta = swapDelta - hookDelta;
+
+            // check that the hook's returns haven't flipped the sign of the tokens being swapped
+            if (
+                params.zeroForOne
+                    ? (swapperDelta.amount0() > 0 || swapperDelta.amount1() < 0)
+                    : (swapperDelta.amount1() > 0 || swapperDelta.amount0() < 0)
+            ) {
+                revert HookDeltaExceedsSwapAmount();
```

Comment:
I think this was the portion we said we didn't mind taking out and having the router check - since this is afterSwap?

Reply:
removed now